hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java

```diff
@@ -192,14 +200,65 @@ private record Validated(@NonNull Account account, @NonNull List<Token> tokens)
     @Override
     public Fees calculateFees(@NonNull final FeeContext feeContext) {
         requireNonNull(feeContext);
+        requireNonNull(feeContext.configuration());
+
         final var body = feeContext.body();
         final var op = body.tokenAssociateOrThrow();
-        final var accountId = op.accountOrThrow();
-        final var readableAccountStore = feeContext.readableStore(ReadableAccountStore.class);
-        final var account = readableAccountStore.getAccountById(accountId);
+        final var calculator = feeContext.feeCalculator(DEFAULT);
+        final var unlimitedAutoAssociations =
+                feeContext.configuration().getConfigData(EntitiesConfig.class).unlimitedAutoAssociationsEnabled();
+
+        if (unlimitedAutoAssociations) {
+            final var exchangeRateInfo = feeContext.exchangeRateInfo();
+
+            calculator.resetUsage();
+            final var singleSigPrice = getTinybarsFromTinyCents(
+                    calculator.getVptPrice(), exchangeRateInfo.activeRate(feeContext.consensusNow()));
+            final var newAssociationsCount = op.tokens().size();
+
+            final var associateInTinyCents = getFixedAssociatePriceInTinyCents();
+            final var associateInTinybars = getTinybarsFromTinyCents(
+                    associateInTinyCents, exchangeRateInfo.activeRate(feeContext.consensusNow()));
+
+            var totalSigPrice = 0L;
+            // If the transaction category is USER then this is the main transaction. If it is CHILD
+            // then this is a transaction that was internally dispatched and the signature verifications
+            // were already performed and charged in the main transaction so we don't need to
+            // calculate them again here.
+            if (feeContext.transactionCategory() == TransactionCategory.USER) {
+                totalSigPrice = Math.max(0, feeContext.numTxnSignatures() - 1) * singleSigPrice;
+            }
+            return new Fees(0L, totalSigPrice + newAssociationsCount * associateInTinybars, 0L);
```

Comment:
This turns the fee calculation upside-down. You should take a `FeeCalculator`, provide the used resources (e.g. number of signatures), and call `calculate()` to get the value.

Reply:
Since the token associate price is now fixed, in the handler we needed to have the canonical price for an association and the price for any additional signatures apart from the first one that is free. Any other calculations in the calculator and the `calculate()` method needed to be reset as they increased additionally the result price. We have discussed this with @tinker-michaelj and @Neeharika-Sompalli and we used the `TokenMintHandler` as reference. This is why needed to expose the `getVptPrice()` method in `FeeCalculator`.

Reply:
Yes @netopyr discussions converged on wanting fees to be a very exact `$0.05`-per-association; but with an increased fee at current prices for each additional signature.

Hacking resource prices in _feeSchedules.json_ to get this behavior would not be a **huge** amount of work, but would:
  1. Make it less obvious what is going on; and,
  2. Require a council-signed update to file `0.0.111`.

We can however, do this, and up somewhere that is more like the standard pattern.

Reply:
Yes, please do this.

Fee calculation is the core responsibility of the `FeeCalculator`. We have to strictly enforce this to avoid a mess. If the current functionality is insufficient, we should discuss whether the `FeeCalculator` can be extended. This will result in a cleaner design than implementing exceptional cases directly in the handlers.

Reply:
In general, keeping the design clean is always a much higher priority than implementing a temporary feature flag.

Why do you think the `feeSchedules.json` approach is problematic? We can upload it anytime, as we would toggle a feature flag.

Reply:
We can have different `feeSchedules.json` files depending on the environment.

Reply:
@netopyr, we opened a separate PR with the refactored version of the code - https://github.com/hashgraph/hedera-services/pull/13942. Can you please review it and let us know if the new approach is better?